### PR TITLE
[SPARK-31552][SQL] Fix ClassCastException in ScalaReflection arrayClassFor

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

the 2 method `arrayClassFor` and `dataTypeFor` in `ScalaReflection` call each other circularly, the cases in `dataTypeFor` are not fully handled in `arrayClassFor`

For example:
```scala
scala> implicit def newArrayEncoder[T <: Array[_] : TypeTag]: Encoder[T] = ExpressionEncoder()
newArrayEncoder: [T <: Array[_]](implicit evidence$1: reflect.runtime.universe.TypeTag[T])org.apache.spark.sql.Encoder[T]

scala> val decOne = Decimal(1, 38, 18)
decOne: org.apache.spark.sql.types.Decimal = 1E-18

scala> val decTwo = Decimal(2, 38, 18)
decTwo: org.apache.spark.sql.types.Decimal = 2E-18

scala> val decSpark = Array(decOne, decTwo)
decSpark: Array[org.apache.spark.sql.types.Decimal] = Array(1E-18, 2E-18)

scala> Seq(decSpark).toDF()
java.lang.ClassCastException: org.apache.spark.sql.types.DecimalType cannot be cast to org.apache.spark.sql.types.ObjectType
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$arrayClassFor$1(ScalaReflection.scala:131)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:879)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:878)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:49)
  at org.apache.spark.sql.catalyst.ScalaReflection$.arrayClassFor(ScalaReflection.scala:120)
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$dataTypeFor$1(ScalaReflection.scala:105)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:879)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:878)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:49)
  at org.apache.spark.sql.catalyst.ScalaReflection$.dataTypeFor(ScalaReflection.scala:88)
  at org.apache.spark.sql.catalyst.ScalaReflection$.$anonfun$serializerForType$1(ScalaReflection.scala:399)
  at scala.reflect.internal.tpe.TypeConstraints$UndoLog.undo(TypeConstraints.scala:69)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects(ScalaReflection.scala:879)
  at org.apache.spark.sql.catalyst.ScalaReflection.cleanUpReflectionObjects$(ScalaReflection.scala:878)
  at org.apache.spark.sql.catalyst.ScalaReflection$.cleanUpReflectionObjects(ScalaReflection.scala:49)
  at org.apache.spark.sql.catalyst.ScalaReflection$.serializerForType(ScalaReflection.scala:393)
  at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder$.apply(ExpressionEncoder.scala:57)
  at newArrayEncoder(<console>:57)
  ... 53 elided

scala>
```

In this PR, we add the missing cases to `arrayClassFor`

### Why are the changes needed?

bugfix as described above

### Does this PR introduce any user-facing change?

no

### How was this patch tested?

add a test for array encoders

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc: @cloud-fan @dongjoon-hyun @HyukjinKwon @maropu thanks

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -2358,6 +2360,79 @@ class DataFrameSuite extends QueryTest
...
+    // are not supported by array
+    implicit def newArrayEncoder[T <: Array[_] : TypeTag]: Encoder[T] = ExpressionEncoder()
+
+    // decimals
+    val decSchema = new StructType()
+      .add("value", "array<decimal(38,18)>")
+    val decSpark = Array(decOne, decTwo)
+    val decScala = decSpark.map(_.toBigDecimal)
+    val decJava = decSpark.map(_.toJavaBigDecimal)
+    checkAnswer(Seq(decSpark).toDF(), new GenericRowWithSchema(Array(decJava), decSchema))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2405]: Why we need `GenericRowWithSchema` here?

Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: I used this for debugging, change it back to Row now

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 2]
```diff
@@ -126,6 +126,9 @@ object ScalaReflection extends ScalaReflection {
       case t if isSubtype(t, definitions.ShortTpe) => classOf[Array[Short]]
       case t if isSubtype(t, definitions.ByteTpe) => classOf[Array[Byte]]
       case t if isSubtype(t, definitions.BooleanTpe) => classOf[Array[Boolean]]
+      case t if isSubtype(t, localTypeOf[Array[Byte]]) => classOf[Array[Array[Byte]]]
+      case t if isSubtype(t, localTypeOf[CalendarInterval]) => classOf[Array[CalendarInterval]]
+      case t if isSubtype(t, localTypeOf[Decimal]) => classOf[Array[Decimal]]
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:131]: Thanks. It seems that this matched with `dataTypeFor` now.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 3]
```diff
@@ -23,26 +23,28 @@ import java.sql.{Date, Timestamp}
...
 import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanHelper
 import org.apache.spark.sql.execution.aggregate.HashAggregateExec
 import org.apache.spark.sql.execution.exchange.{BroadcastExchangeExec, ReusedExchangeExec, ShuffleExchangeExec}
 import org.apache.spark.sql.functions._
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.test.{ExamplePoint, ExamplePointUDT, SharedSparkSession}
-import org.apache.spark.sql.test.SQLTestData.{DecimalData, NullStrings, TestData2}
-import org.apache.spark.sql.types._
+import org.apache.spark.sql.test.SQLTestData.{DecimalData, TestData2}
+import org.apache.spark.sql.types.{Decimal, _}
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:47]: The original line 45 seems to be enough because it was `import org.apache.spark.sql.types._`. Can we revert this change?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 4]
```diff
@@ -2358,6 +2360,74 @@ class DataFrameSuite extends QueryTest
     val df = Seq((1, new CalendarInterval(1, 2, 3))).toDF("a", "b")
     checkAnswer(df.selectExpr("b"), Row(new CalendarInterval(1, 2, 3)))
   }
+
+  test("array encoder with different types") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2364]: Shall we add `SPARK-31552: ` prefix? This is a bug for all old Spark versions. For backporting to branch-2.4, it would be helpful.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 5]
```diff
@@ -2358,6 +2360,74 @@ class DataFrameSuite extends QueryTest
...
+    checkAnswer(Seq(strings).toDF(), Row(strings))
+
+    // tuples
+    val decOne = Decimal(1, 38, 18)
+    val decTwo = Decimal(2, 38, 18)
+    val tuple1 = (1, 2.2, "3.33", decOne, Date.valueOf("2012-11-22"))
+    val tuple2 = (2, 3.3, "4.44", decTwo, Date.valueOf("2022-11-22"))
+    checkAnswer(Seq(Array(tuple1, tuple2)).toDF(), Seq(Seq(tuple1, tuple2)).toDF())
+
+    // TODO: we can move this implicit def to [[SQLImplicits]] when we eventually make fully
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2393]: @yaooqinn . Since you are an active contributor, I want to give you a comment on this. Since Apache Spark repo is a shared repository, you had better use an IDed TODO like `TODO(SPARK-XXX)` after making a JIRA issue.

However, in this case, I guess you just wanted to add a comment about the possibility. So, in this case, please remove `TODO:` simply.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 6]
```diff
@@ -2358,6 +2360,74 @@ class DataFrameSuite extends QueryTest
...
+    checkAnswer(Seq(instants).toDF(), Row(timestamps))
+
+    val intervals = Array(new CalendarInterval(1, 2, 3), new CalendarInterval(4, 5, 6))
+    checkAnswer(Seq(intervals).toDF(), Row(intervals))
+
+    // binary
+    val bins = Array(Array(1.toByte), Array(2.toByte), Array(3.toByte), Array(4.toByte))
+    checkAnswer(Seq(bins).toDF(), Row(bins))
+
+    // nested
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2427]: Can we use a nested decimal array here?

**Comment by dongjoon-hyun** [comment]
Thank you for fixing this, @yaooqinn and @cloud-fan .

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 7]
```diff
@@ -2358,6 +2360,74 @@ class DataFrameSuite extends QueryTest
...
+
+    val strings = Array("2020-04-24", "2020-04-25")
+    checkAnswer(Seq(strings).toDF(), Row(strings))
+
+    // tuples
+    val decOne = Decimal(1, 38, 18)
+    val decTwo = Decimal(2, 38, 18)
+    val tuple1 = (1, 2.2, "3.33", decOne, Date.valueOf("2012-11-22"))
+    val tuple2 = (2, 3.3, "4.44", decTwo, Date.valueOf("2022-11-22"))
+    checkAnswer(Seq(Array(tuple1, tuple2)).toDF(), Seq(Seq(tuple1, tuple2)).toDF())
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2391]: Just a question: We didn't have a test coverage for these primitive types and tuples?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 5 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2393]: OK, roger that

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 6 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2431]: just add one more decimal case here

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @yaooqinn and @cloud-fan .
Merged to master/3.0.

**Comment by dongjoon-hyun** [comment]
Could you make a backporting PR to `branch-2.4`, @yaooqinn ?

**Comment by maropu** [comment]
late LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 8]
```diff
@@ -126,6 +126,9 @@ object ScalaReflection extends ScalaReflection {
       case t if isSubtype(t, definitions.ShortTpe) => classOf[Array[Short]]
       case t if isSubtype(t, definitions.ByteTpe) => classOf[Array[Byte]]
       case t if isSubtype(t, definitions.BooleanTpe) => classOf[Array[Boolean]]
+      case t if isSubtype(t, localTypeOf[Array[Byte]]) => classOf[Array[Array[Byte]]]
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala:129]: nit: might be worth updating the comment above: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala#L117-L118

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 9]
```diff
@@ -2358,6 +2360,80 @@ class DataFrameSuite extends QueryTest
...
+
+    // case classes
+    val gbks = Array(GroupByKey(1, 2), GroupByKey(4, 5))
+    checkAnswer(Seq(gbks).toDF(), Row(Array(Row(1, 2), Row(4, 5))))
+
+    // We can move this implicit def to [[SQLImplicits]] when we eventually make fully
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:2397]: let's don't `[[...]]` or unlined comments. That syntax is only valid for proper multiline docs in Scaladoc. You can just do `` `...` `` next time.

**Review by HyukjinKwon — COMMENTED** [review]
LGTM too
